app/src/test/java/android/util/Log.java

```diff
@@ -0,0 +1,35 @@
+package android.util;
+
+/**
+ * Workaround slf4j trying to invoke Android.Log.isLoggable
+ * by providing a shim which simply uses println on everything.
+ */
+public class Log {
```

Comment:
The rest of the project uses Timber for logging. Why use System.out instead?

Reply:
So there's a bit of a hidden chain going on. Tl;dr anything that uses logs (such as Mockk or tested code) will invariantly require Android classes that aren't available in JVM only tests.

The longer explanation:
`Timber` and `SLF4J` are logging frameworks, which indirectly pipe to a logging API (usually the Android system). As `Mockk` will emit debug logging to allow you to track invocations / usage of the mocks this also pulls in the logging frameworks. Unit testing code which logs could potentially do the same thing too (but as of this PR nothing we test touches logs).
As this is a JVM unit test (as opposed to running on the Android Emulator) all the `Android.*` imports are unavailable and will produce the following error:
```
Caused by: java.lang.RuntimeException: Method isLoggable in android.util.Log not mocked. See http://g.co/androidstudio/not-mocked for details.
	at android.util.Log.isLoggable(Log.java)
	at uk.uuid.slf4j.android.LogAdapter.getNativeLogLevel(LogAdapter.java:113)
	at uk.uuid.slf4j.android.LogAdapter.<init>(LogAdapter.java:65)
	at uk.uuid.slf4j.android.LoggerFactory.<clinit>(LoggerFactory.java:46)
	... 50 more
```

So ultimately we're left with 3 options:
- Run the unit test with the Android emulator, just to put the logging API we will ignore in
- Before each test, mock out `Android.utils.Log` to make the framework happy (but this has a chicken and egg problem too so we'd need something like Powermockito)
- Have our own implementation of `Android.utils.Log` that only exists for JVM test targets that simply forwards to `stdout` as per this (SO post)[https://stackoverflow.com/a/46793567]

If these were integration tests I'd opt for the former. But this would still prevent us writing unit tests which (IMO) are a nice form of test for people to write and something is better than nothing in my mind.

I hope that explains the rationale, and I'm happy to hear your thoughts on this?

Reply:
I'll mark this as resolved, as the offending commit is now primarily in #1312